File: c++/Maximum Average Pass Ratio.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <utility>
#include <iomanip>

class Solution {
public:
    double maxAverageRatio(std::vector<std::vector<int>>& classes, int extraStudents) {
        auto compare = [](const std::pair<int, int>& a, const std::pair<int, int>& b) {
            double gainA = ((double)a.first + 1) / (a.second + 1) - (double)a.first / a.second;
            double gainB = ((double)b.first + 1) / (b.second + 1) - (double)b.first / b.second;
            return gainA < gainB; 
        };

        std::priority_queue<std::pair<int, int>, std::vector<std::pair<int, int>>, decltype(compare)> maxHeap(compare);

        for (const auto& c : classes) {
            maxHeap.push({c[0], c[1]});
        }

        while (extraStudents-- > 0) {
            auto [pass, total] = maxHeap.top();
            maxHeap.pop();
            maxHeap.push({pass + 1, total + 1});
        }

        double totalAverage = 0.0;
        while (!maxHeap.empty()) {
            auto [pass, total] = maxHeap.top();
            maxHeap.pop();
            totalAverage += (double)pass / total;
        }

        return totalAverage / classes.size();
    }
};
// Test cases
int main() {
    Solution solution;
  
    // Test case 1
    std::vector<std::vector<int>> classes1 = {{1, 2}, {3, 5}, {2, 2}};
    int extraStudents1 = 2;
    double result1 = solution.maxAverageRatio(classes1, extraStudents1);
    std::cout << std::fixed << std::setprecision(5) << result1 << "\n"; // Expected output: 0.78333

    // Test case 2
    std::vector<std::vector<int>> classes2 = {{2, 4}, {3, 9}, {4, 5}, {2, 10}};
    int extraStudents2 = 4;
    double result2 = solution.maxAverageRatio(classes2, extraStudents2);
    std::cout << std::fixed << std::setprecision(5) << result2 << "\n"; // Expected output: 0.53485

    return 0;
}

File: python/Maximum Average Pass Ratio.py
import heapq
from typing import List

class Solution:
    def maxAverageRatio(self, classes: List[List[int]], extraStudents: int) -> float:
        """
        Distribute extra students across classes to maximize the average pass ratio.
        """
        
        def gain(p: int, t: int) -> float:
            return (p + 1) / (t + 1) - p / t

        # Max-heap simulated by storing negative gains
        heap = [(-gain(p, t), p, t) for p, t in classes]
        heapq.heapify(heap)

        for _ in range(extraStudents):
            g, p, t = heapq.heappop(heap)
            p, t = p + 1, t + 1
            heapq.heappush(heap, (-gain(p, t), p, t))

        total_average = sum(p / t for _, p, t in heap)
        return total_average / len(classes)

# Test cases
if __name__ == "__main__":
    sol = Solution()

    # case 1
    classes1 = [[1, 2], [3, 5], [2, 2]]
    extraStudents1 = 2
    print(f"Result of test case 1 : {sol.maxAverageRatio(classes1, extraStudents1)}")  # Expected ~0.78333

    # case 2
    classes2 = [[2, 4], [3, 9], [4, 5], [2, 10]]
    extraStudents2 = 4
    print(f"Result of test case 2 : {sol.maxAverageRatio(classes2, extraStudents2)}")  # Expected ~0.53485

